### ISSUE TITLE
decide how to handle the breakout of the dockershim from the kubelet.

### ISSUE THREAD

**Issue by neolit123** [kubernetes/kubeadm#1412]
this issue tracks the dockershim deprecation / removal with some key events over time.

**summary** of the kubeadm stance for the dockershim removal:
- if only docker is present kubeadm will try to use it but a 1.24> kubelet will fail.
you'd have to build and deploy cri-dockerd on the host that exposes the dockershim CRI socket:
https://github.com/Mirantis/cri-dockerd
- if you have feedback about the dockershim removal please comment on https://github.com/kubernetes/kubernetes/issues/106917
- if you have multiple container runtime sockets on the host kubeadm will throw an error asking you to pick one
this is possible with init|joinconfiguration.noderegistration.crisocket
- if you only have one socket - containerd or crio, kubeadm will autodetect and use the socket
- docs / migration guides from docker to another container runtime will contain details about kubeadm
- kubeadm no longer has e2e tests for docker as the CR. in the future we can have tests if cri-dockerd becomes maintained and has docs / releases.

**history:**

update 19.05.2020:
- building kubelet without dockershim is now possible:
https://github.com/kubernetes/enhancements/pull/1546
- next step is breaking out dockershim (new KEP)?

update 11.11.2020:
- dockershim has been deprecated in 1.20:
https://github.com/kubernetes/kubernetes/pull/94624
https://github.com/kubernetes/enhancements/pull/1985

update 1.24 / 12.2021:
- dockershim was removed. more details below:
https://github.com/kubernetes/kubeadm/issues/1412#issuecomment-976777974
- kubeadm no longer tests (until further notice) with docker as the CR in e2e tests:
https://github.com/kubernetes/kubeadm/pull/2620
- dockershim flags are removed:
https://github.com/kubernetes/kubernetes/pull/106907

**Comment by rosti** [comment]
I'll move in on this one.

/assign

**Comment by timothysc** [comment]
We're blocked on kubelet changes first.

**Comment by astrieanna** [comment]
/remove-help

This does not fit the [help wanted definition](https://github.com/kubernetes/community/blob/master/contributors/guide/help-wanted.md) because:
* It seems to involve community discussion/decisions (e.g. how to distribute the binary).
* It seems to be taken already (by @rosti).
* It is blocked (by kubelet changes).

**Comment by BenTheElder** [comment]
Should cri-dockerd now be distributed by mirantis etc. downstream just as docker and containerd etc. are handled by their vendor(s) / distros? (given that the kubernetes project is not going to own this binary)

**Comment by neolit123** [comment]
my vote goes for not releasing cri-dockerd as part of the k8s release packages, though this is up to sig-release to decide.
we should probably stop doing the same for crictl too. the current version is outdated but our packages depend on it:
https://github.com/kubernetes/release/issues/1636
and a wild guess is that if we change this it will break apt/rpm users.

on the kubeadm side, my prior proposal was to stop managing the CR cgroup driver and sandbox image for all CRs. docker was the only one we did that for and now that it becomes external we should stop doing these actions completely.

a few documentation updates will be required for users to prepare these during node setup at:
https://kubernetes.io/docs/setup/production-environment/container-runtimes/

**Comment by xlgao-zju** [comment]
@neolit123 if the docker-shim is remove by kubelet, how will we handle the upgrade of control-plane which is using docker as runtime?

**Comment by neolit123** [comment]
The plan is for dockershim to be a separete binary downloaded from a third
party location. It would become the responsibility of the cluster admin to
install / configure / upgrade it.

**Comment by dims** [comment]
please see https://github.com/Mirantis/cri-dockerd

**Comment by neolit123** [comment]
looks like dockershim was removed from the kubelet in k/k master
https://github.com/kubernetes/kubernetes/pull/97252

this means we have to switch our e2e jobs against k/k master to containerd. since our setup is rather complex it would be easier to switch the jobs against all k/k branches to containerd.

to do that we have to modify the base image in all the files here:
https://github.com/kubernetes/kubeadm/blob/ecb3d2206a00b1aab25d754a0c4c95d781ac0909/kinder/ci/tools/update-workflows/templates/workflows/
to be `baseImage: kindest/base:v20191105-ee880e9b`

and then run ./hack/update-workflows.sh in kinder.

**Comment by neolit123** [comment]
one problem in the kubeadm code base is here:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/kubelet/flags.go#L80-L86

with our support skew we'd have to not manage this branching for kubelet N, while we have to manage it for kubelet N-1.

i think this issue can be closed after that.

in an ideal world the https://github.com/Mirantis/cri-dockerd project should include docs on how to deploy a service (e.g. managed by systemd) for users that never want to move away from docker(shim).also  assuming the project will gain maintenance traction.

**Comment by sftim** [comment]
BTW, is the description of this issue current? (if not, would it be useful to revise it?)

**Comment by neolit123** [comment]
good point, updated with a summary

**Comment by neolit123** [comment]
cross linking to the tracking issue for dockershim removal feedback:
https://github.com/kubernetes/kubernetes/issues/106917

**Comment by afbjorklund** [comment]
> in an ideal world the https://github.com/Mirantis/cri-dockerd project should include docs on how to deploy a service (e.g. managed by systemd) for users that never want to move away from docker(shim).also assuming the project will gain maintenance traction.

Currently they refer to https://docs.mirantis.com, which doesn't have _any_ information for cri-dockerd

**Comment by neolit123** [comment]
closing in favor of remain tasks for 1.24 and later TBD here:
https://github.com/kubernetes/kubeadm/issues/2626
